runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaPublishFactory.java

```diff
@@ -375,76 +380,82 @@ private void onMqttData(
-            {
-                final String16FW responseTopic = mqttPublishDataEx.responseTopic();
-                addHeader(helper.kafkaReplyToHeaderName, messages.topic);
-                addHeader(helper.kafkaReplyKeyHeaderName, responseTopic);
+                if (mqttPublishDataEx.responseTopic().length() != -1)
+                {
+                    final String16FW responseTopic = mqttPublishDataEx.responseTopic();
+                    addHeader(helper.kafkaReplyToHeaderName, messages.topic);
+                    addHeader(helper.kafkaReplyKeyHeaderName, responseTopic);
 
-                addFiltersHeader(responseTopic);
-            }
+                    addFiltersHeader(responseTopic);
+                }
 
-            if (mqttPublishDataEx.correlation().bytes() != null)
-            {
-                addHeader(helper.kafkaCorrelationHeaderName, mqttPublishDataEx.correlation().bytes());
-            }
+                if (mqttPublishDataEx.correlation().bytes() != null)
+                {
+                    addHeader(helper.kafkaCorrelationHeaderName, mqttPublishDataEx.correlation().bytes());
+                }
 
 
-            mqttPublishDataEx.properties().forEach(property ->
-                addHeader(property.key(), property.value()));
+                mqttPublishDataEx.properties().forEach(property ->
+                    addHeader(property.key(), property.value()));
 
-            addHeader(helper.kafkaQosHeaderName, qosLevels.get(mqttPublishDataEx.qos()));
+                addHeader(helper.kafkaQosHeaderName, qosLevels.get(mqttPublishDataEx.qos()));
 
-            final int deferred = mqttPublishDataEx.deferred();
-            kafkaDataEx = kafkaDataExRW
-                .wrap(extBuffer, 0, extBuffer.capacity())
-                .typeId(kafkaTypeId)
-                .merged(m -> m.produce(mp -> mp
-                    .deferred(deferred)
-                    .timestamp(now().toEpochMilli())
-                    .partition(p -> p.partitionId(-1).partitionOffset(-1))
-                    .key(b -> b.set(key))
-                    .hashKey(this::setHashKey)
-                    .headers(kafkaHeadersRW.build())))
-                .build();
+                final int deferred = mqttPublishDataEx.deferred();
+                kafkaDataEx = kafkaDataExRW
+                    .wrap(extBuffer, 0, extBuffer.capacity())
+                    .typeId(kafkaTypeId)
+                    .merged(m -> m.produce(mp -> mp
+                        .deferred(deferred)
+                        .timestamp(now().toEpochMilli())
+                        .partition(p -> p.partitionId(-1).partitionOffset(-1))
+                        .key(b -> b.set(key))
+                        .hashKey(this::setHashKey)
+                        .headers(kafkaHeadersRW.build())))
+                    .build();
 
+                retainedData = (mqttPublishDataEx.flags() & 1 << MqttPublishFlags.RETAIN.value()) != 0;
```

Comment:
Please hoist `1 << MqttPublishFlags.RETAIN.value()` as `PUBLISH_FLAGS_RETAINED_MASK`.
Please remember `publishFlags` instead of `retainedData` during fragmented message publish.
Please add a static method `boolean hasPublishFlagRetained(int publishFlags)` which is used for `if (retainedData)` condition below.

This will help the readability and better convey the intent I think.